File: mersenne.py
from functools import lru_cache


@lru_cache(maxsize=100000000)
def isPrime(num):
    for i in range(2, num):
        if num % i == 0:
            return False
    return True


@lru_cache(maxsize=100000000)
def primes(array):
    prime = []
    indexes = []
    for i in range(0, len(array)):
        if isPrime(array[i]):
            prime.append(array[i])
            indexes.append(array.index(array[i]))
    indexes = tuple(indexes)
    prime = tuple(prime)
    return (prime, indexes)


@lru_cache(maxsize=100000000)
def mersennes(n):
    pArray = []
    for i in range(0, n):
        pArray.append(i)
    pArray = tuple(pArray)
    primed = primes(pArray)
    primed = primed[0]
    primed = tuple(primed)
    # print(primed)
    possibleMersenne = []
    mersenneFor = []
    for i in range(0, len(primed)):
        mersenneFor.append('2^' + str(primed[i]) + ' - 1')
        possibleMersenne.append((2**primed[i]) - 1)
    possibleMersenne = tuple(possibleMersenne)
    mersenne = primes(possibleMersenne)
    corrFor = []
    for i in range(0, len(mersenne[1])):
        corrFor.append(mersenneFor[mersenne[1][i]])
    both = {}
    for i in range(0, len(corrFor)):
        both[corrFor[i]] = mersenne[0][i]
    return both


retrn = mersennes(100)
print(retrn)

File: mersenne.js
const inp = require('readline-sync');
let isPrime = num => {
    for (var i = 2; i < num; i++) {
        if (num % i === 0) {
            return false;
        }
    }
    return true;
}
let primes = array => {
    var primes = []
    var indexes = []
    for (let i = 1; i < array.length; i++) {
        if (isPrime(array[i])) {
            primes.push(array[i])
            indexes.push(array.indexOf(array[i]));
        }
    }
    return [primes, indexes];
};
let mersenne = limit => {
    var pArray = [];
    for (let p = 0; p < limit; p++) {
        pArray.push(p);
    }
    var prime = primes(pArray)[0];
    var possibleMersennes = [];
    var mersenneFor = []
    for (let i = 0; i < prime.length; i++) {
        mersenneFor.push(`2^${prime[i]} - 1`)
        possibleMersennes.push(Math.pow(2, prime[i]) - 1);
    }
    // console.log(possibleMersennes);
    var mersenne = primes(possibleMersennes);
    var corrFor = [];
    for (let i = 0; i < mersenne[1].length; i++) {
        corrFor.push(mersenneFor[mersenne[1][i]]);
    }
    // console.log(mersenne[0]);
    // console.log(mersenne[1]);
    // console.log(corrFor);
    var mersennes = {};
    for (let k = 0; k < corrFor.length; k++) {
        mersennes[corrFor[k]] = mersenne[0][k];

    }
    return mersennes;
};
var retrn = mersenne(100);
console.log(retrn);
// console.log(retrn.length);